src/main/java/org/prebid/server/bidder/criteo/CriteoBidder.java

```diff
@@ -0,0 +1,279 @@
+                    .distinct()
+                    .count() > 1) {
+                return Result.withError(
+                      BidderError.badInput("Bid request has slots coming with several network IDs which is not allowed")
+                );
+            }
+
+            criteoRequestBuilder.slots(criteoRequestSlots);
+        }
+
+        final Regs regs = bidRequest.getRegs();
+        final ExtRegs extRegs = regs != null ? regs.getExt() : null;
+
+        criteoRequestBuilder
+                .publisher(buildCriteoPublisher(bidRequest, networkId))
+                .user(buildCriteoUser(bidRequest, extRegs))
+                .gdprConsent(buildCriteoGdprConsent(bidRequest, extRegs));
+
+        if (bidRequest.getUser() != null && bidRequest.getUser().getExt() != null) {
+            ExtUser extUser = bidRequest.getUser().getExt();
+            criteoRequestBuilder.eids(extUser.getEids());
+        }
+
+        final CriteoRequest criteoRequest = criteoRequestBuilder.build();
+        final String requestBody;
+        try {
+            requestBody = jsonMapper.encode(criteoRequest);
+        } catch (EncodeException e) {
+            return Result.withError(BidderError.badInput(
+                    String.format("Failed to encode request body, error: %s", e.getMessage())));
+        }
+
+        return Result.withValue(HttpRequest.<CriteoRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .body(requestBody)
+                .headers(resolveHeaders(criteoRequest))
+                .payload(criteoRequest)
+                .build());
+    }
+
+    private ExtImpCriteo parseImpExt(Imp imp) {
+        try {
+            return jsonMapper.mapper().convertValue(imp.getExt(), CRITEO_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private CriteoPublisher buildCriteoPublisher(BidRequest bidRequest, Integer networkId) {
+        CriteoPublisher.CriteoPublisherBuilder criteoPublisherBuilder = CriteoPublisher.builder();
+        if (networkId != null) {
+            criteoPublisherBuilder.networkId(networkId);
+        }
+
+        if (bidRequest.getApp() != null) {
+            criteoPublisherBuilder.bundleId(bidRequest.getApp().getBundle());
+        }
+
+        if (bidRequest.getSite() != null) {
+            criteoPublisherBuilder
+                    .siteId(bidRequest.getSite().getId())
+                    .url(bidRequest.getSite().getPage());
+        }
+
+        return criteoPublisherBuilder.build();
+    }
+
+    private CriteoUser buildCriteoUser(BidRequest bidRequest, ExtRegs extRegs) {
+        CriteoUser.CriteoUserBuilder criteoUserBuilder = CriteoUser.builder();
+        if (bidRequest.getUser() != null) {
+            criteoUserBuilder.cookieId(bidRequest.getUser().getBuyeruid());
+        }
+
+        if (bidRequest.getDevice() != null) {
```

Comment:
Probably worth declaring `final Device device = bidRequest.getDevice()` and reuse `device` variable below